### DISCUSSION TITLE
Generate a truly empty solution

### DISCUSSION THREAD

**Discussion by mobynet1** [unoplatform/uno#17744] [Q&A] [answered]
### What would you like to be added

I have spent close to 18 years developing WPF applications. I have tried hard to make Maui work, but alas, even after two-plus years, it is still not ready and I'm doubtful it ever will be.

I have struggled to generate a purely empty Uno application. I have reported three issues, [#17732](https://github.com/unoplatform/uno/issues/17732), [#17733](https://github.com/unoplatform/uno/issues/17733), and [#17734](https://github.com/unoplatform/uno/issues/17734), which are the results of trying to get a truly empty solution upon which to build my knowledge. I find it to be of no help getting an install with a populated App.xaml/App.xaml.cs file and the Main and Second pages because there is no clear way of how these are constructed. Even looking at the tutorials does not make it clear.

What I *would* find helpful is a stepwise tutorial on how to gen up a new Uno solution from scratch. First, you would explain what goes into App.xaml and why. Then you would explain how to add a blank first page and what to add to it to get the app up and running. I have always developed apps using XAML so making the switch to C# Markup is something I have wanted to do because I agree that it makes the items that were once developed in XAML more compact and provide more control of how conditions affect the flow.

1. Being able to use full conditional logic that affects styling would be straightforward as opposed to the limited "is equal" and "is not equal" which can only be assembled with AND logic. I don't know how many times I have wanted to use OR logic in my triggers. Yes, complex multi-triggers can achieve this. Converts can aid. But the effort one has to go through to make a cell turn red based on a set of conditions is ridiculous. C# Markup will simplify this tremendously.
2. Being able to see the anatomy of a configuration from app to main to sub-pages would make it easier to see how things fit together.

Once the basics are understood, it becomes easy to simply create a page and the code-behind logic and begin being productive far more rapidly than having template code provided which one takes for granted without truly understanding the linkage between all the moving parts.

### Why is this needed

I just need my learning to be facilitated by snapping the Lego blocks together myself to see how things are assembled.

### For which platform

WebAssembly, Android, iOS, Mac Catalyst, Skia (Linux X11), Skia (macOS), Skia (Linux Framebuffer), Skia (GTK)

**Comment by mobynet1** [comment]
It was not supposed to be opened as an issue... Don't now how it happened...

**Reply by jeromelaban** [reply]
Thanks for your reports. What you've opened is more of a discussion than an issue at this point. If we determine pieces that need to be fixed, we'll open appropriate issues.

I'll split the conversations in order for making it easier to follow.

**Comment by mobynet1** [comment]
UPDATE: I tried to create an "empty" solution by manually entering the switches to the `dotnet new unoapp` command and got close... I issued this command

```text
dotnet new unoapp --name MobyClient --output MobyClientUno --language="C#" --preset blank --publisher "MobyNet Corp." --app-id "io.mobynet.MobyClient" -tfm net8.0 -presentation mvux -markup "csharp" --app-theme material -tests none -toolkit --localization true -cultures="ar|cs|da|nl|fr|de|es|it|ja|ko|pt-BR|ru|sv|tr|uk|vi|zh-Hans|zh-Hant" --logging none --navigation blank -theme-service true --unoSvg true -dsp true --continuous-integration none -rider false
```

but it complained about the list of cultures. I tried many variations on the switch without success. I tried bars ("|"). I tried commas. I tried enclosing each culture in quotes: "ar","cs". I tried -cultures=*list* ... and nothing worked.

I did get this command to work:

```text
dotnet new unoapp --name MobyClient --output MobyClientUno --language="C#" --preset blank --publisher "MobyNet Corp." --app-id "io.mobynet.MobyClient" -tfm net8.0 -presentation mvux -markup "csharp" --app-theme material -tests none -toolkit --localization true --logging none --navigation blank -theme-service true --unoSvg true -dsp true --continuous-integration none -rider false
```

What was wrong with the original command?

**Reply by jeromelaban** [reply]
> UPDATE: I tried to create an "empty" solution by manually entering the switches to the dotnet new unoapp command and got close... I issued this command

I'm not sure what you're referring to and why you need to create the `dotnet new` manually. Which parameters are you trying to use that are not in the wizard?

**Reply by mobynet1** [reply]
I was just trying to see how the `dotnet new` created the solution and what the various switches did. It was for my learning as it were...

**Comment by jeromelaban** [comment]
> I have struggled to generate a purely empty Uno application. I have reported three issues, https://github.com/unoplatform/uno/issues/17732, https://github.com/unoplatform/uno/issues/17733, and https://github.com/unoplatform/uno/issues/17734, which are the results of trying to get a truly empty solution upon which to build my knowledge. I find it to be of no help getting an install with a populated App.xaml/App.xaml.cs file and the Main and Second pages because there is no clear way of how these are constructed. Even looking at the tutorials does not make it clear.

It's important to know that Uno Platform is based off of WinUI, and you can find everything you need for navigation and pages [there](https://learn.microsoft.com/en-us/windows/apps/design/basics/navigate-between-two-pages) or [there](https://learn.microsoft.com/en-us/windows/apps/design/controls/navigationview).

> What I would find helpful is a stepwise tutorial on how to gen up a new Uno solution from scratch. First, you would explain what goes into App.xaml and why. Then you would explain how to add a blank first page and what to add to it to get the app up and running. I have always developed apps using XAML so making the switch to C# Markup is something I have wanted to do because I agree that it makes the items that were once developed in XAML more compact and provide more control of how conditions affect the flow.

This is pretty close to the same as above. The APIs are ones coming from WinUI and we're relying on Microsoft's documentation where it makes sense. Now, I understand that having a basic walkthrough of the app would be of interest and knowing what you're having difficulties grasping in the current structure would be interesting to have in this particular [solution structure page](https://platform.uno/docs/articles/uno-app-solution-structure.html). Could you tell us more?

**Comment by jeromelaban** [comment]
> Being able to see the anatomy of a configuration from app to main to sub-pages would make it easier to see how things fit together.

Could you give an example of what you're looking for?

**Comment by mobynet1** [comment]
When I developed WPF applications, I know there was a Learn.Microsoft.com page somewhere where it broke the windows structure down and related it to components in the apps definition files. I looked for it, but could not find it. Some things I did find were items like this: [Overview of WPF windows](https://learn.microsoft.com/en-us/dotnet/desktop/wpf/windows/?view=netdesktop-6.0) and this: [Desktop Guide](https://learn.microsoft.com/en-us/dotnet/desktop/wpf/overview/?view=netdesktop-8.0). I realize this only deals with WPF on Windows only and is not cross-platform at all, but the diagram I remember broke a window, its chrome, and its components down and showed how those things could be overridden by creating custom XAML, but this would be invaluable for all of the platforms. For instance, when implement something for macOS, how can you affect the status bar at the top of each app and what has to be done to affect it. If you want to do "X", then change "this" in the common code, or change "this" in the platform code and where in the code to do it.

I am looking at how the different styles are added to the Resources.Build command to merge various dictionaries together. Knowing the "internals" of a step-wise recipe that would say these are all the minimum things that belong in the App.xaml.cs file, in what order, if that is critical, and what are some of the optional things that can be added and what they would affect.

When I created custom chrome for my windows, I had to do "this" to the app.xaml file, "this" in the mainwindow.xaml file, and "this" in the mainwindowviewmodel.cs to get a foundation upon which to build the app. I used to use whatever the first attempt to decouple all of the components in WPF prior to Prism. How to set up regions, navigation, etc. The brain cell where the name of this first framework (SCC-something-or-other) is long gone, but the Prism docs went into great detail in documenting all of the moving parts. And documenting all of this requires a ton of examples. I detest looking at documentation for an API and it gives a one line, lame sentence of what it is. Duh! That is easy. When I first started developing software back in 1984, it was on a VAX/VMS 750, but the doc set for VMS (it used to be called the Orange Wall, then DEC changed the color of the 20 or 30 three inch three ring binders to grey) was phenomenal and I have never seen any documentation like it since. Their Run Time Library binder describe each API in excruciating detail and gave detailed examples in C of how to use each API. While it would take an army of technical writers a year to do that for Uno and I would not expect it, going from a trivial explanation where nothing can be gleaned from it to a masterpiece where you could OCR the document and develop a system would be the Holy Grail.

Of course, us silly end-users expect the world and you might be able to provide a small patch of dirt because that is all you could provide with the resources you have, but you see where I am going with this. You should see my code. I document EVERYTHING because someday I want to retire, but I want my applications to live on!! The only way for that to happen is to leave a map for the next guy who sits in my chair!

I can see why this was moved to a discussion! ;-)

**Reply by francoistanguay** [reply]
I'd be happy to have a longer conversation on this but thought I'd share some initial thoughts:

First, Uno's promise is to "bridge" or implement WinUI"s API on other platforms. So a good place to start when wanting to understand the  lifecycle and capabilities of the framework can be to look into WinUI's (and previously UWP) extensive documentation, samples, tutorials, community content.

Second, we tend to focus our documentation on what's specific to Uno.

Third, our main focus is to make sure you can get up and running quickly and see a sneak peak of everything Uno has to offer (Themes, Toolkit, Extensions for Navigation, Authentication, ... C# Markup, Hot Reload, IDE Extensions, ...) and that's why when you use the Wizard, by default, it uses the "recommended" template that brings quite a few things with it. Obviously you can select the blank template and opt-in to whatever you're interested in.

Fourth, and in all cases, you did find friction points we totally need to eliminate. It might be because you hit an edge case (we have so many combinations of platforms, IDE and packages that it's extremely difficult to test them all) or because a dependency we have might have broken a portion of the experience recently (looking at you C# Dev Kit).

Finally, we appreciate the feedback, your patience and the requested details you might be able to provide. Can't promise it'll be perfect but can promise we'll be responsive and do our best as an open-source project to eliminate all the friction we can uncover.

**Reply by mobynet1** [reply, marked as answer]
@francoistanguay Thanks for this response. I congratulate you on the Uno Platform overall. It has a steep learning curve, but I am making it work!

Through trial and error, I have created many new projects just to see how they're configured based on the options I selected in the solution wizard. I feel like I have come to understand the process and now have a working solution that fulfills my desires. I am off and running and am liking what I am seeing! Good work Uno Team!

**Reply by francoistanguay** [reply]
Thank you for the support and feedback! As you encounter new issues, please continue to file them, open discussions, share repros... THIS is how we're benefiting the most from the community. The better the details and repros, the faster we can commit on fixing them.

**Reply by DevTKSS** [reply]
@mobynet1 I am also just started to learn uno, with no xaml or .net core/Standard pre-knowledge, only vb.net in framework 🙈 
You did say, that you made progress and understand... I am still feeling kind of helpless and overwelmed, not knowing when to do what at all in this workflow. I would really appreciate if you could maybe tell me, what was your steps to get to the "Ahh! Now that makes all sense!" Point I am still missing on my mind. Please let me know.

**Reply by francoistanguay** [reply]
Where are you struggling? 

Have you looked into our Getting Started? https://platform.uno/docs/articles/get-started.html?tabs=windows

Have you gone through our Tutorials/Samples? https://platform.uno/docs/articles/samples-tutorials-overview.html

**Reply by DevTKSS** [reply]
@francoistanguay
Yes, I did the get started and also did the counter app. Tryed the TubePlayer but I do not have a google cloud membership so that is not useable to me althought that would maybe clear some questionmarks for my plan for the solution I want to finally create at some point with uno app, but the very beginning is like a huge wall right infront of me, not knowing how to climb it.
To this a really broken down guide like he described in the post above would maybe help on that... a guide that is not only telling to write that code down, more explaining what this is actually doing and the possible arguments...

I did see the samples, yes. And after about 3 Weeks I also did recognize, that uno.samples repo including a UI folder which is actually containing interesting samples and not only the reference folder which is containing the language and pattern differenced ones.

What I also do not really understand on this samples:
![image](https://github.com/user-attachments/assets/d899266c-fd45-4b50-a109-43779c70cbbc)
e.g. here is this line that says "Download the complete sample"
but I would really not know how to do that? This is shown to me, when clicking this link 🤔 
![image](https://github.com/user-attachments/assets/85a3574c-bdbc-4934-a796-59174739a826)

And as I said, I did find the samples, which are listed on the documentation, thats right, but to most of them, only the repo source is linked without any future description. So actually my thought has also been, is it possible to actually download just a specific solution from this samples and how would that would look like? I would prefer to not download the very whole uno.samples repo, to be honest 😟 My computer has not infinit storage for telling the truth...

For example, Topic that I dont understand also after reading the documantation and clicking throught the documentation folders to maybe see a hidden page to that like the navigation How-to really is to me...
- Navigation, really biiiiig questionmark and unresolvable to me, because it is telling about the how to navigate in xaml and between pages(views) but beside the NavigateViewAsync or similar, I do not understand the NavigateData or NavigateResult things and the NavigateMessageDialogAsyc? As beginner I thought, the "recommended" template would be a good idea, but the How-To on this:
https://platform.uno/docs/articles/external/uno.extensions/doc/Learn/Navigation/HowTo-NavigateBetweenPages.html which is telling initially, it is for the recommended template, using mvux, is just switching to mvvm from the 5. element, since there is suddenly a file named SampleViewModel reffered to edit, which is not in our hands on mvux.
So I did want to show a "simple" (thought it would be simple) ContentDialog in a c# mvux app and try out the navigation thing of uno:
https://platform.uno/docs/articles/external/uno.extensions/doc/Learn/Navigation/HowTo-NavigateBetweenPages.html
Which is (to me) not clear from this, because all UI definition here is with xaml. Did list the enchancement idea of c# template for this already but at this time I think the team has enough to do beside such small things 😄 but having a template for this for a c# markup app withough xaml might would make it possible for a beginner like me to use and understand it?
on the lower part of the linked page is the modal contentdialog I wanted to create, but again only xaml markup, no Idea how to get it in c# markup. also checked out the gallery to this, but seems like there is no mvux example, only xaml with xaml.cs, because of the routedeventargs thing there.
So, my try for mvux and c# Contentdialog failed on this #18657 so as always I am touching software (also as user XD) the bugs seems to hunt me...

And here something where the dog is biting its own tail: 
https://platform.uno/docs/articles/features/dialogs.html?tabs=mvvm%2Cxaml
beside the fact, that there is the IXamlRootProvider mentioned which is no where decribed (add somewhere? intellisense did not gues a namespace I could import for this) If you scroll to the bottom of that Page, you will find this:
![image](https://github.com/user-attachments/assets/b9a42b5e-1813-461d-b6b2-dcf67bc07f07)
I did mark the ??? area for you. So as you can see, it is clearly telling to please not use MessageDialog in Uno Applications.
But then, while reading the Navigation How-To Pages, I did stumble over this page:
https://platform.uno/docs/articles/external/uno.extensions/doc/Learn/Navigation/HowTo-DisplayMessageDialog.html
and my mind is still running circles on that. Please, please 🙏 could someone clarify if this decision to use MessageDialog is "use" or "dont use"? With telling "no" in the actual ContentDialog Chapter and "Yes" in the Navigator Chapter, that is driving me crazy 😅
It would not make sense to write a how-to page for a thing that should not be used 🤔 ❓

So e.g. on this example topic my only intention was a simple ContentDialog and all I did find were workarounds...
